tests/go/go-kafka-serde/client.go

```diff
@@ -267,13 +251,116 @@ func (tc *TestClient) RunTest() error {
 			}
 
 			numConsumed++
-		case kafka.Error:
-			util.Die("Failure during consumption: %v", e)
+		case *kafka.Error:
+			return e
 		default:
 			log.Debugf("Ignoring unknown message: %v", e)
 		}
 	}
 
+	if numConsumed != tc.count {
+		util.Die("Did not consume expected number of messages")
+	}
+
+	return nil
+}
+
+func (tc *TestClient) produceMessages() error {
+	numDelivered := 0
+
+	// Create a context with a cancel function
+	produceCheckContext, cancelProduceCheck := context.WithCancel(context.Background())
+	// Create an error group that will be used to monitor for cancellation and
+	// message production events from franz-go producer
+	produceCheckGroup, produceCheckContext := errgroup.WithContext(produceCheckContext)
+
+	monitorProduceFn := func(ctx context.Context) func() error {
+		return func() error {
+			for {
+				select {
+				case <-ctx.Done():
+					log.Debugf("Monitor Produce received done signal: %v", ctx.Err())
+					return ctx.Err()
+				case e := <-tc.producer.Events():
+					switch ev := e.(type) {
+					case *kafka.Message:
+						m := ev
+
+						if m.TopicPartition.Error != nil {
+							log.Errorf("Failed to produce to topic %v: %v", tc.topic, m.TopicPartition.Error)
+						} else {
+							numDelivered++
+
+							if numDelivered == tc.count {
+								log.Infof("Produced expected number of messages")
+								return nil
+							}
+						}
+					case kafka.Error:
+						log.Errorf("Failed to produce to topic %v: %v", tc.topic, ev)
+						return ev
+					default:
+						log.Debugf("Ignoring event %v", ev)
+					}
+				}
+			}
+		}
+	}
+
+	produceCheckGroup.Go(monitorProduceFn(produceCheckContext))
+
+	defer cancelProduceCheck()
+
+	for i := 0; i < tc.count; i++ {
+		payload_value, err := tc.serializer.CreateSerializedData(i, &tc.topic)
+		if err != nil {
+			return err
+		}
+		log.Debugf("payload: %v", payload_value)
+		err = tc.producer.Produce(&kafka.Message{
+			TopicPartition: kafka.TopicPartition{Topic: &tc.topic, Partition: kafka.PartitionAny},
+			Value:          payload_value,
+		}, nil)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create a new context that will timeout after specified timeout
+	timeoutContext, timeoutCancel := context.WithTimeout(produceCheckContext, time.Duration(tc.timeout)*time.Second)
+	// Create an error group that will monitor for the timeout event
+	timeoutGroup, timeoutContext := errgroup.WithContext(timeoutContext)
```

Comment:
Why do we need a group here for a single thing that doesn't error?

Reply:
Definitely open to suggestions on how to make this better!  This is my first attempt at Go :)  

Namely just need to capture a timeout to cancel the producer check context when it expires.

Reply:
Here's a little toy that is maybe simpler: https://go.dev/play/p/QVBUDriTkEy

I think you can just drop the second group and launch a goroutine that is cancellable.

Not sure if that's helpful, but I think you want something like this:

- Setup monitor producer for recording events (goroutine)
- Produce all the messages
- Setup a timeout for the producer to finish

I think you only need a single WaitGroup to make sure that monitor and timeout goroutines finish, then you can have two contexts and let each goroutine cancel out each other.

I think a code smell here is that timeoutGroup is never waited upon, and it's only used as cancellation.